packages/openapi3/src/types.ts

```diff
@@ -39,6 +40,57 @@ export interface OpenAPI3Document extends Extensions {
   security?: Record<string, string[]>[];
 }
 
+/**
+ * A record containing the the OpenAPI 3 documents corresponding to
+ * a particular service definition.
+ */
+export type OpenAPI3ServiceRecord =
+  | OpenAPI3UnversionedServiceRecord
+  | OpenAPI3VersionedServiceRecord;
+
+export interface OpenAPI3UnversionedServiceRecord {
+  /** The service that generated this OpenAPI document */
+  service: Service;
+
+  /** Whether the service is versioned */
+  versioned: false;
+
+  /** The OpenAPI 3 document */
+  document: OpenAPI3Document;
+
+  /** The diagnostics created for this document */
+  diagnostics: Readonly<Diagnostic[]>;
+}
+
+export interface OpenAPI3VersionedServiceRecord {
+  /** The service that generated this OpenAPI document */
+  service: Service;
+
+  /** Whether the service is versioned */
+  versioned: true;
+
+  /** The OpenAPI 3 document records for each version of this service */
+  versions: OpenAPI3VersionedDocumentRecord[];
+}
+
+/**
+ * A record containing an unversioned OpenAPI document and associated metadata.
+ */
+
+export interface OpenAPI3VersionedDocumentRecord {
+  /** The OpenAPI document*/
+  document: OpenAPI3Document;
+
+  /** The service that generated this OpenAPI document. */
+  service: Service;
+
+  /** The version of the service. Absent if the service is unversioned. */
+  version: string;
+
+  /** The diagnostics created for this version. */
+  readonly diagnostics: Readonly<Diagnostic[]>;
```

Comment:
```suggestion
  readonly diagnostics: readonly Diagnostic[];
```

also should all the props be readonly?